services/secondary_signals_updater.py

```diff
@@ -0,0 +1,187 @@
+    this field is maintained by ATD staff via the signals forms in the knack database
+
+    Args:
+        signals (TYPE): Description
+
+    Returns:
+        TYPE: Description
+    """
+    signals_with_children = {}
+
+    for signal in signals:
+
+        try:
+            #  get id of parent signal
+            knack_id = signal[field_maps["PRIMARY_SIGNAL"]][0]["id"]
+        except (IndexError, AttributeError):
+            #  empty key
+            continue
+
+        if knack_id not in signals_with_children:
+            #  add entry for parent signal
+            signals_with_children[knack_id] = []
+        #  add current signal to list of parent's children
+        signals_with_children[knack_id].append(signal["id"])
+
+    return signals_with_children
+
+
+def get_old_prim_signals(signals, field_maps):
+    """
+    create a dict of primary signals and the secondary signals they control
+    data is compiled from the 'secondary_signals' field on primary signals
+    this field is populated by this Python service
+
+    Args:
+        signals (TYPE): Description
+
+    Returns:
+        TYPE: Description
+    """
+    signals_with_children = {}
+
+    for signal in signals:
+        knack_id = signal["id"]
+
+        secondary_signals = []
+
+        try:
+            for secondary in signal[field_maps["SECONDARY_SIGNALS"]]:
+                secondary_signals.append(secondary["id"])
+
+                signals_with_children[knack_id] = secondary_signals
+
+        except (KeyError, AttributeError):
+            continue
+
+    return signals_with_children
+
+
+def main(args):
+    # Parse Arguments
+    app_name = args.app_name
+    container = args.container
+    logger.info(args)
+
+    # Selecting correct config for the view
+    config = CONFIG[app_name][container]
+    field_mapping = SECONDARY_SIGNALS[app_name][container]
+
+    # Get Signals Knack Data
+    kwargs = {"scene": config["scene"], "view": container}
+    data = knackpy.api.get(
+        app_id=APP_ID, api_key=API_KEY, **kwargs
+    )
+
+    primary_signals_old = get_old_prim_signals(data, field_mapping)
+    primary_signals_new = get_new_prim_signals(data, field_mapping)
+
+    """
+    Checking for three separate cases where we need to update Knack
+    1. A new secondary signal was added to a primary signal
+    2. Secondary signal(s) was/were changed that are attached a primary signal
+    3. A secondary signal was removed from a primary signal 
+    """
+
+    payload = []
+
+    for signal_id in primary_signals_new:
+        """
+        identify new and changed primary-secondary relationships
+        """
+        if signal_id in primary_signals_old:
+            new_secondaries = collections.Counter(primary_signals_new[signal_id])
+            old_secondaries = collections.Counter(primary_signals_old[signal_id])
+
+            # Checking if things have changed
+            if old_secondaries != new_secondaries:
```

Comment:
i had forgotten that you can do dict equality in python!

```python
{"`a": "b", "c": "d"} == {"c": "d", "a": "b"}
```

Reply:
this is your handy work!

Reply:
♾️